dcos_launch/terraform.py

```diff
@@ -0,0 +1,285 @@
+            os.chmod(self.terraform_binary, 0o100)
+        finally:
+            os.remove(download_path)
+        log.info('Terraform installation complete.')
+
+    def wait(self):
+        """ Nothing to do here because unlike the other launchers, create() runs and also waits for a subprocess to
+        finish instead of just sending an http request to a provider.
+        """
+        pass
+
+    def delete(self):
+        # delete the cluster
+        subprocess.run([self.terraform_cmd(), 'destroy', '-force', '-var-file', self.cluster_profile_path],
+                       cwd=self.init_dir, check=True, stderr=subprocess.STDOUT, env=os.environ)
+        # remove the init dir
+        shutil.rmtree(self.init_dir, ignore_errors=True)
+
+    def describe(self) -> dict:
+        """ Sample output from 'terraform output' command:
+        Bootstrap Host Public IP = 35.227.147.39
+        Master ELB Public IP = 35.230.64.74
+        Master Public IPs = [
+            35.230.56.247
+        ]
+        Private Agent Public IPs = [
+            35.230.42.164,
+            35.203.136.227
+        ]
+        Public Agent ELB Public IP = 35.230.71.57
+        Public Agent Public IPs = [
+            35.230.88.105
+        ]
+        ssh_user = core
+        """
+        result = subprocess.run([self.terraform_cmd(), 'output'], cwd=self.init_dir, check=True, stdout=subprocess.PIPE)
+        info = result.stdout.decode('utf-8')
+        self.config['ssh_user'] = re.search('ssh_user = \w+', info).group(0)[len('ssh_user = '):]
+
+        private_agents_ips = _convert_to_describe_format(_get_ips('Private Agent Public IPs', info))
+        private_agents_gpu_addresses = _get_ips('GPU Public IPs', info)
+        for i in range(len(private_agents_gpu_addresses)):
+            private_agents_ips[i]['GPU Public IPs'] = private_agents_gpu_addresses[i]
+
+        description = {
+            'bootstrap_host': _convert_to_describe_format(_get_ips('Bootstrap Host Public IP', info=info)),
+            'masters': _convert_to_describe_format(_get_ips('Master Public IPs', info=info)),
+            'private_agents': private_agents_ips,
+            'public_agents': _convert_to_describe_format(_get_ips('Public Agent Public IPs', info))}
+
+        master_elb_address = _get_ips('Master ELB Public IP', info=info)
+        public_agent_elb_address = _get_ips('Public Agent ELB Public IP', info=info)
+
+        if master_elb_address:
+            description.update({'Master ELB Public IP': master_elb_address[0]})
+        if public_agent_elb_address:
+            description.update({'Public Agent ELB Public IP': public_agent_elb_address[0]})
+
+        return description
+
+    def _key_helper_common(self, private_key: bytes):
+        self.config['ssh_private_key'] = private_key.decode('utf-8')
+        with open(self.default_priv_key_path, 'wb') as f:
+            f.write(private_key)
+        os.chmod(self.default_priv_key_path, 0o600)
```

Comment:
I would say the same point on octal codes as above, but SSH is so picky that its still going to be obtuse with the abstractions